pkg/controller/operators/olm/apiservices.go

```diff
@@ -128,16 +129,69 @@ func (a *Operator) checkAPIServiceResources(csv *v1alpha1.ClusterServiceVersion,
 			}
 		}
 
-		// Ensure the existing deployment has a matching CA hash annotation
+		// Ensure the existing Deployment has a matching CA hash annotation
 		deployment, err := a.lister.AppsV1().DeploymentLister().Deployments(csv.GetNamespace()).Get(desc.DeploymentName)
 		if k8serrors.IsNotFound(err) || err != nil {
-			logger.Warnf("expected APIService Deployment %s could not be retrieved", desc.DeploymentName)
+			logger.Warnf("expected Deployment %s could not be retrieved", desc.DeploymentName)
 			return err
 		}
-		if hash, ok := deployment.GetAnnotations()[OLMCAHashAnnotationKey]; !ok || hash != caHash {
+		if hash, ok := deployment.Spec.Template.GetAnnotations()[OLMCAHashAnnotationKey]; !ok || hash != caHash {
 			logger.Warnf("Deployment %s CA cert hash does not match expected", desc.DeploymentName)
 			return fmt.Errorf("Deployment %s CA cert hash does not match expected", desc.DeploymentName)
 		}
+
+		// Ensure the Deployment's ServiceAccount exists
+		serviceAccountName := deployment.Spec.Template.Spec.ServiceAccountName
+		serviceAccount, err := a.lister.CoreV1().ServiceAccountLister().ServiceAccounts(deployment.GetNamespace()).Get(serviceAccountName)
+		if err != nil {
+			logger.Warnf("could not retrieve ServiceAccount %s", serviceAccountName)
+			return err
+		}
+
+		// Ensure RBAC permissions for the APIService are correct
+		rulesMap := map[string][]rbacv1.PolicyRule{
+			// Serving cert Secret Rule
+			csv.GetNamespace(): {
+				{
+					Verbs:         []string{"get"},
+					APIGroups:     []string{""},
+					Resources:     []string{"secrets"},
+					ResourceNames: []string{secret.GetName()},
+				},
+			},
+			"kube-system":       {},
+			metav1.NamespaceAll: {},
+		}
+
+		// extension-apiserver-authentication-reader
+		authReaderRole, err := a.lister.RbacV1().RoleLister().Roles("kube-system").Get("extension-apiserver-authentication-reader")
+		if err != nil {
+			logger.Warnf("could not retrieve Role extension-apiserver-authentication-reader")
+			return err
+		}
+		rulesMap["kube-system"] = append(rulesMap["kube-system"], authReaderRole.Rules...)
+
+		// system:auth-delegator
+		authDelegatorClusterRole, err := a.lister.RbacV1().ClusterRoleLister().Get("system:auth-delegator")
+		if err != nil {
+			logger.Warnf("could not retrieve ClusterRole system:auth-delegator")
+			return err
+		}
+		rulesMap[metav1.NamespaceAll] = append(rulesMap[metav1.NamespaceAll], authDelegatorClusterRole.Rules...)
+
+		for namespace, rules := range rulesMap {
```

Comment:
Is checking if the required rules are satisfied better than just checking if the generated Roles/Bindings exist by name? I'm wondering if we can also get into an infinite CSV transition (Installing -> Failed -> Pending -> Installing -> Failed) because OLM doesn't have permissions to create these resources. I think the answer is no, but would like a second pair of eyes to validate that.

Reply:
Definitely better to check the rules when we can. 

You bring up a good point, we ideally only want to transition back to pending _if_ we have reason to believe that the final result could change (e.g. previously rules were not met, but now they are, or vice-versa).

I think we probably want to address this by recording on the status each succeess/failure check, as a "real" Condition (as opposed to our Conditions which are just a history of transitions). Then there could be a "RulesMet" condition, a "ComponentHealth" condition (or many of them), etc. When we're in success/failure, we want to detect if those things have changed in a way that would move the needle, and then flip back to pending.

I'm not sure we need to do this here (we do have a story for Conditions) unless you think that's the only way to reasonably avoid the loops you're talking about. I think it should be enough to carry this discussion into the design docs for Conditions and work on it later.